rust/rbac-registration/src/cardano/cip509/validation.rs

```diff
@@ -1,679 +1,409 @@
-                            "{function_name}, Failed to get index of payment key: {e}"
-                        ));
-                        return None;
-                    },
-                };
-                let outputs = tx.transaction_body.outputs.clone();
-                let witness = match TxWitness::new(&[txn.clone()]) {
-                    Ok(witnesses) => witnesses,
-                    Err(e) => {
-                        validation_report
-                            .push(format!("{function_name}, Failed to create TxWitness: {e}"));
-                        return None;
-                    },
-                };
-
-                if let Some(output) = outputs.get(index) {
-                    match output {
-                        pallas::ledger::primitives::conway::PseudoTransactionOutput::Legacy(o) => {
-                            return validate_payment_output_key_helper(
-                                &o.address.to_vec(),
-                                validation_report,
-                                &witness,
-                            );
-                        },
-                        pallas::ledger::primitives::conway::PseudoTransactionOutput::PostAlonzo(
-                            o,
-                        ) => {
-                            return validate_payment_output_key_helper(
-                                &o.address.to_vec(),
-                                validation_report,
-                                &witness,
-                            );
-                        },
-                    };
+        if !matches!(metadata.x509_certs.first(), Some(X509DerCert::X509Cert(_)))
+            && !matches!(
+                metadata.c509_certs.first(),
+                Some(C509Cert::C509Certificate(_))
+            )
+        {
+            report.other("The role 0 certificate must be present", context);
+        }
+    } else {
+        // For other roles there still must be exactly one certificate at 0 index, but it must
+        // have the `undefined` value.
+        if matches!(metadata.x509_certs.first(), Some(X509DerCert::X509Cert(_)))
+            || matches!(
+                metadata.c509_certs.first(),
+                Some(C509Cert::C509Certificate(_))
+            )
+        {
+            report.other("Only role 0 can contain a certificate at 0 index", context);
+        }
+        if matches!(metadata.x509_certs.first(), Some(X509DerCert::Deleted))
+            || matches!(metadata.c509_certs.first(), Some(C509Cert::Deleted))
+        {
+            report.other("Only role 0 can delete a certificate at 0 index", context);
+        }
+    }
+
+    // It isn't allowed for any role to use a public key at 0 index.
+    if !matches!(
+        metadata.pub_keys.first(),
+        None | Some(SimplePublicKeyType::Undefined)
+    ) {
+        report.other(
+            "The public key cannot be used for the role 0, only a certificate",
```

Comment:
Since this is checking for other role too. Maybe better to say
"Role 0 requires a certificate, not a public key; for other roles, set the first public key to undefined if needed."

Or add this check for role 0 and other role so the error messages are distinct